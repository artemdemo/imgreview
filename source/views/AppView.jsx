import React from 'react';
import { connect } from 'react-redux';
import _get from 'lodash/get';
import MainMenu from '../components/MainMenu/MainMenu';
import AppVersion from '../components/AppVersion/AppVersion';
import CanvasContainer from '../containers/CanvasContainer/CanvasContainer.async';
import MIOpenImage from '../containers/MenuItems/MIOpenImage.async';
import MISave from '../containers/MenuItems/MISave';
import MIArrow from '../containers/MenuItems/MIArrow.async';
import MIGithub from '../containers/MenuItems/MIGithub';
import MIStrokeColor from '../containers/MenuItems/MIStrokeColor';
import MIResize from '../containers/MenuItems/MIResize/MIResize';
import {
    blurShapes,
} from '../model/shapes/shapesActions';

class AppView extends React.PureComponent {
    componentDidMount() {
        document.addEventListener('click', this.clickOnBody);
    }

    componentWillUnmount() {
        document.removeEventListener('click', this.clickOnBody);
    }

    clickOnBody = (e) => {
        const isHtmlEl = _get(e.target, 'tagName') === 'HTML';
        const isDivApp = e.target && e.target.getAttribute('id') === 'app';
        if (isHtmlEl || isDivApp) {
            const { blurShapes } = this.props;
            blurShapes();
        }
    };

    onMenuClick = () => {
        const { blurShapes } = this.props;
        blurShapes();
    };

    render() {
        return (
            <React.Fragment>
                <AppVersion />
                <MainMenu
                    onClick={this.onMenuClick}
                >
                    <MIOpenImage />
                    <MISave />
                    <MIArrow />
                    <MIStrokeColor />
                    <MIResize />
                    <MIGithub />
                </MainMenu>
                <CanvasContainer />
            </React.Fragment>
        );
    }
}

export default connect(
    () => ({}), {
        blurShapes,
    },
)(AppView);
