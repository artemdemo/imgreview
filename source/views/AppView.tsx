import React from 'react';
import _get from 'lodash/get';
import AppVersion from '../components/AppVersion/AppVersion';
import CanvasContainer from '../containers/CanvasContainer/CanvasContainer.async';
import Menu from '../containers/Menu/Menu';
import { TStateCanvas } from '../model/canvas/canvasReducer';
import * as canvasApi from '../canvas/api';

import '../styles/general.less';

type Props = {
    canvas: TStateCanvas;
};

class AppView extends React.PureComponent<Props> {
    componentDidMount() {
        document.addEventListener('click', this.clickOnBody);
    }

    componentWillUnmount() {
        document.removeEventListener('click', this.clickOnBody);
    }

    clickOnBody = (e) => {
        const isHtmlEl = _get(e.target, 'tagName') === 'HTML';
        const isDivApp = e.target && e.target.getAttribute('id') === 'app';
        if (isHtmlEl || isDivApp) {
            canvasApi.blurShapes();
        }
    };

    render() {
        return (
            <React.Fragment>
                <AppVersion />
                <Menu />
                <CanvasContainer />
            </React.Fragment>
        );
    }
}

export default AppView;
