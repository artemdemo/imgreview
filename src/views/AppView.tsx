import React, { useContext, useEffect } from 'react';
import _ from 'lodash';
import { AppVersion } from '../components/AppVersion/AppVersion';
import { Menu } from '../containers/Menu/Menu';
import { MobileWarning } from '../components/MobileWarning/MobileWarning';
import { Suspense } from '../components/Suspense/Suspense';
import { AppStateContext } from '../model/AppStateContext';
import '../styles/general.less';

const Notifications = React.lazy(
  () =>
    import(
      /* webpackChunkName: "Notifications" */
      '../containers/Notifications/Notifications'
    ),
);

const CanvasContainer = React.lazy(
  () =>
    import(
      /* webpackChunkName: "CanvasContainer" */
      '../containers/CanvasContainer/CanvasContainer'
    ),
);

export const AppView: React.FC = () => {
  const {
    state: {
      menu,
      canvas: { canvasApi },
    },
  } = useContext(AppStateContext);

  useEffect(() => {
    const clickOnBody = (e: any) => {
      const isHtmlEl = _.get(e.target, 'tagName') === 'HTML';
      const isDivApp = e.target && e.target.getAttribute('id') === 'app';
      if (isHtmlEl || isDivApp) {
        canvasApi?.blurShapes();
      }
    };
    document.addEventListener('click', clickOnBody);
    return () => {
      document.removeEventListener('click', clickOnBody);
    };
  }, [ canvasApi ]);

  return (
    <>
      <AppVersion />
      <Menu />
      <Suspense
        fallback={
          <div style={{ position: 'absolute', top: menu.menuHeight }}>
            Loading...
          </div>
        }
      >
        <CanvasContainer />
      </Suspense>
      <Suspense>
        <Notifications />
      </Suspense>
      <MobileWarning />
    </>
  );
};
