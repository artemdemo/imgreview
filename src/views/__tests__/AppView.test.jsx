import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import AppView from '../AppView';

jest.mock('../../../srcCanvas/api');
jest.mock('../../components/TopMenu/TopMenuPanel');
jest.mock('../../components/AppVersion/AppVersion');
jest.mock('../../containers/CanvasContainer/CanvasContainer.async');
jest.mock('../../containers/Menu/Menu');

describe('AppView', () => {
  const canvasApiMock = require('../../../srcCanvas/api');
  const addEventListenerSpy = jest.spyOn(document, 'addEventListener');
  const removeEventListenerSpy = jest.spyOn(document, 'removeEventListener');

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render', () => {
    const tree = renderer.create(<AppView />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should mount and unmount', () => {
    const wrapper = mount(<AppView />);

    const instance = wrapper.instance();
    expect(addEventListenerSpy).toBeCalledWith('click', instance.clickOnBody);
    wrapper.unmount();
    expect(removeEventListenerSpy).toBeCalledWith(
      'click',
      instance.clickOnBody
    );
  });

  describe('clickOnBody', () => {
    it('should call if id == "app"', () => {
      const wrapper = mount(<AppView />);
      const instance = wrapper.instance();
      const getAttributeMock = jest.fn(() => 'app');
      instance.clickOnBody({
        target: {
          getAttribute: getAttributeMock,
        },
      });
      expect(canvasApiMock.blurShapes).toBeCalled();
      expect(getAttributeMock).toBeCalledWith('id');
    });

    it('should call if target is HTML element', () => {
      const wrapper = mount(<AppView />);
      const instance = wrapper.instance();
      instance.clickOnBody({
        target: {
          tagName: 'HTML',
          getAttribute: () => {},
        },
      });
      expect(canvasApiMock.blurShapes).toBeCalled();
    });

    it('should not call clickOnBody', () => {
      const wrapper = mount(<AppView />);
      const instance = wrapper.instance();
      instance.clickOnBody({});
      expect(canvasApiMock.blurShapes).not.toBeCalled();
    });
  });
});
