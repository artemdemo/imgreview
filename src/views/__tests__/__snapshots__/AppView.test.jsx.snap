// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppView should render 1`] = `
Array [
  <div
    data-mock="AppVersion"
  />,
  <div
    data-mock="Menu"
  />,
  <div
    data-mock="CanvasContainer"
  />,
]
`;
